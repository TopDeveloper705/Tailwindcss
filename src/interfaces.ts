import type { Utility } from './lib/utilities/handler';

import type { Property, Style, InlineAtRule } from './utils/style';

export type DictStr = { [key: string]: string };

export type DeepNestDictStr = { [key:string]: string | DeepNestDictStr };

export type NestObject = { [key: string]: string | string[] |NestObject };

export type DeepNestObject = { [key: string]: NestObject };

export type GenericNestObject<T> = { [key: string]: T | GenericNestObject<T> };

export type AnyObject = Record<string, unknown>;

export type AnyValue<T> = T;

export type Output = Property | Style | Style[] | undefined;

export type FontSize = [
  fontSize?: string,
  options?: { letterSpacing?: string; lineHeight?: string }
];

export type DefaultFontSize = [
  fontSize: string,
  options: { letterSpacing?: string; lineHeight: string }
];

export type ThemeUtil = (path: string, defaultValue?: unknown) => unknown;

export type ConfigUtil = (
  theme: ThemeUtil,
  {
    negative,
    breakpoints,
  }: {
    negative: (config: DictStr) => DictStr;
    breakpoints: (config: DictStr) => DictStr;
  }
) => unknown;

export type PluginUtilOptions =
  | string[]
  | {
      variants?: string[];
      respectPrefix?: boolean;
      respectImportant?: boolean;
    };

export interface PluginBuilder {
  (handler: (utils: PluginUtils) => void, config?: Config): PluginOutput;
  withOptions: <T = DictStr>(
    pluginFunction: (options: T) => ((utils: PluginUtils) => void),
    configFunction?: (options: T) => Config,
  ) => PluginWithOptions<T>;
}

export interface PluginOutput {
  handler: (utils: PluginUtils) => void;
  config?: Config;
  __isOptionsFunction?: false;
}

export interface PluginWithOptions<T = DictStr>{
  (options?: T): PluginOutputWithOptions<T>
  __isOptionsFunction: true;
  __pluginFunction: (options: T) => ((utils: PluginUtils) => void);
  __configFunction: (options: T) => Config;
}

export interface PluginOutputWithOptions<T = DictStr> extends PluginOutput {
  __options: T;
}

export interface Theme {
  [key: string]: ConfigUtil | { [ key:string ]: unknown } | undefined;
}

export type Plugin = PluginOutput | PluginWithOptions<unknown> | PluginOutputWithOptions<unknown>

export interface Config {
  presets?: Config[];
  separator?: string;
  important?: boolean | string;
  darkMode?: 'media' | 'class' | false;
  theme?: Theme;
  variantOrder?: string[];
  plugins?: Plugin[];
  corePlugins?: string[];
  prefix?: string;

  /**
   * @deprecated no longer needed for Windi CSS
   */
  purge?: unknown
  /**
   * @deprecated no longer needed for Windi CSS
   */
  variants?: { [key: string]: string[] };
}

export interface DefaultTheme {
  colors: { [key: string]: string | { [key: string]: string } };
  container: { [key: string]: string | { [key: string]: string } };
  fontFamily: { [key: string]: string[] };
  fontSize: { [key: string]: DefaultFontSize };
  keyframes: { [key: string]: { [key: string]: string } };
  outline: { [key: string]: [outline: string, outlineOffset: string] };
}

export interface DefaultTheme {
  [key: string]:
    | DictStr
    | { [key: string]: string | { [key: string]: string } }
    | { [key: string]: string[] }
    | { [key: string]: DefaultFontSize };
}

export interface DefaultConfig {
  presets: string[];
  darkMode: 'class' | 'media' | false;
  theme: DefaultTheme;
  variantOrder: string[];
  variants: { [key: string]: string[] };
  plugins: Plugin[];
}

export interface StaticUtility { [key: string]: { [key: string]: string | string[] } }

export interface DynamicUtility { [key: string]: (utility: Utility, { theme }: PluginUtils) => Output }

export interface PluginUtils {
  addDynamic: (key: string, generator: UtilityGenerator, options?: PluginUtilOptions) => UtilityGenerator;
  addUtilities: (
    utilities: DeepNestObject,
    options?: PluginUtilOptions
  ) => Style[];
  addComponents: (
    components: DeepNestObject | DeepNestObject[],
    options?: PluginUtilOptions
  ) => Style[];
  addBase: (baseStyles: DeepNestObject) => Style[];
  addVariant: (
    name: string,
    generator: VariantGenerator,
  ) => Style | Style[];
  e: (selector: string) => string;
  prefix: (selector: string) => string;
  theme: (path: string, defaultValue?: unknown) => unknown;
  variants: (path: string, defaultValue?: string[]) => string[];
  config: (path: string, defaultValue?: unknown) => unknown;
}

export type VariantGenerator = (generator: {
  modifySelectors: (
    modifier: ({ className }: { className: string }) => string
  ) => Style;
  atRule: (name: string) => Style;
  pseudoClass: (name: string) => Style;
  pseudoElement: (name: string) => Style;
  parent: (name: string) => Style;
  child: (name: string) => Style;
  separator: string;
  style: Style;
}) => Style;

export type UtilityGenerator = (generator: {
  Utility: Utility;
  Style: {
    (
      selector: string,
      property?: Property | Property[] | undefined,
      important?: boolean
    ): Style;
    generate: (
      parent?: string | undefined,
      property?: NestObject | undefined,
      root?: Style | undefined
    ) => Style[];
  };
  Property: {
    (
      name: string | string[],
      value?: string | undefined,
      comment?: string | undefined,
      important?: boolean
    ): Property;
    parse: (
      css: string
    ) => Property | InlineAtRule | (Property | InlineAtRule)[] | undefined;
  };
}) => Output;

export interface Element {
  raw: string;
  start: number;
  end: number;
  variants: string[];
  content?: Element[] | string;
  func?: string;
  type: 'group' | 'func' | 'utility';
}
