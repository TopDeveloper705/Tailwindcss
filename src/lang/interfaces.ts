import type jsep from 'jsep';

export type operand = number | string | boolean;
export type unaryCallback = (a: operand) => operand;
export type binaryCallback = (a: operand, b: operand) => operand;

export type AnyExpression = jsep.ArrayExpression
  | jsep.BinaryExpression
  | jsep.MemberExpression
  | jsep.CallExpression
  | jsep.ConditionalExpression
  | jsep.Identifier
  | jsep.Literal
  | jsep.LogicalExpression
  | jsep.ThisExpression
  | jsep.UnaryExpression;
