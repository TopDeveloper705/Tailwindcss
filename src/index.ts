export * from './processor';
export * from './utils';
export * from './config';
export * from './plugin';

module.exports = {
    'colors': require('./config/colors')
}

