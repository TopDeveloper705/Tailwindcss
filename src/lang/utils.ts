export function isSpace(str: string): boolean {
  return /\s/.test(str);
}

export function isDigit(str: string): boolean {
  return /^\d+$/.test(str);
}

export function isAlpha(str: string): boolean {
  return /\w/.test(str);
}

export function connect(strings: string[]): string {
  return strings.join(';\n') + ';';
}

export function findGroupEnd(text: string, startIndex = 0): number {
  let level = 1;
  let endBracket = searchFrom(text, '}', startIndex);
  while (endBracket !== -1) {
    const nextBracket = searchFrom(text, '{', startIndex);
    if (endBracket < nextBracket || nextBracket === -1) {
      level--;
      startIndex = endBracket + 1;
      if (level === 0) return endBracket;
    } else {
      level++;
      startIndex = nextBracket + 1;
    }
    endBracket = searchFrom(text, '}', startIndex);
  }
  return -1;
}

export function searchFrom(
  text: string,
  target: string | RegExp,
  startIndex = 0,
  endIndex?: number
): number {
  // search from partial of string
  const subText = text.substring(startIndex, endIndex);
  const relativeIndex = subText.search(target);
  return relativeIndex === -1 ? -1 : startIndex + relativeIndex;
}
