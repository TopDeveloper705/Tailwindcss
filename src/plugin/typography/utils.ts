export function uniq<T>(array: T[]): T[] {
  return Array.from(new Set(array));
}

export function castArray<T>(value: T): unknown[] {
  if (!arguments.length) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

export function isUsableColor(color: string, values: string | {[key:string]:string}): boolean {
  return Boolean(values && typeof values === 'object' && color !== 'gray' && values[600]);
}

export const round = (num: number): string =>
  num
    .toFixed(7)
    .replace(/(\.[0-9]+?)0+$/, '$1')
    .replace(/\.0$/, '');

export const rem = (px: number): string => `${round(px / 16)}rem`;

export const em = (px: number, base: number): string => `${round(px / base)}em`;
