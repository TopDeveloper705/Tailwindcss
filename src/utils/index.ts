export * from "./tools";
export * from "./helpers";
