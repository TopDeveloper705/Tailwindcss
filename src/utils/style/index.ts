export { Property, InlineAtRule, Style, GlobalStyle, Keyframes } from './base';
export { StyleSheet } from './sheet';
export * from './prefixer';
