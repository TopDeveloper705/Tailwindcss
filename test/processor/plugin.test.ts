import { FontSize } from '../../dist/types/interfaces';
import { Property, Style } from '../../src/utils/style';
import { Processor } from '../../src/lib';

const processor = new Processor();

describe('Plugin Method', () => {
  it('addUtilities', () => {
    const utilities = {
      '.skew-10deg': {
        transform: 'skewY(-10deg)',
      },
      '.skew-15deg': {
        transform: 'skewY(-15deg)',
      },
    };
    expect(processor.addUtilities(utilities).map(i=>i.build()).join('\n')).toBe('.skew-10deg {\n  transform: skewY(-10deg);\n}\n.skew-15deg {\n  transform: skewY(-15deg);\n}');
  });

  it('addUtilities with non-class styles', () => {
    const utilities = {
      '.skew-10deg': {
        transform: 'skewY(-10deg)',
      },
      '.skew-15deg': {
        transform: 'skewY(-15deg)',
      },
      '[data-theme]': { '--active': '#0099FF' },
    };
    processor.addUtilities(utilities);
    expect(processor.interpret('skew-10deg').styleSheet.build()).toEqual('.skew-10deg {\n  transform: skewY(-10deg);\n}');
    expect(processor.preflight('test', false, false, true).build()).toEqual(
      `[data-theme] {
  --active: #0099FF;
}`);
  });

  it('addComponents', () => {
    const buttons = {
      '.btn': {
        padding: '.5rem 1rem',
        borderRadius: '.25rem',
        fontWeight: '600',
      },
      '.btn-blue': {
        backgroundColor: '#3490dc',
        color: '#fff',
        '&:hover': {
          backgroundColor: '#2779bd',
        },
      },
      '.btn-red': {
        backgroundColor: '#e3342f',
        color: '#fff',
        '&:hover': {
          backgroundColor: '#cc1f1a',
        },
      },
    };

    expect(processor.addComponents(buttons).map(i=>i.build()).join('\n')).toBe('.btn {\n  padding: .5rem 1rem;\n  border-radius: .25rem;\n  font-weight: 600;\n}\n.btn-blue {\n  background-color: #3490dc;\n  color: #fff;\n}\n.btn-blue:hover {\n  background-color: #2779bd;\n}\n.btn-red {\n  background-color: #e3342f;\n  color: #fff;\n}\n.btn-red:hover {\n  background-color: #cc1f1a;\n}');
  });

  it('addComponents with non-class styles ', () => {
    const processor = new Processor();
    const components = {
      '[data-theme]': { '--active': '#0099FF' },
      '[data-theme="dark"]': {
        '--primary': '#111111',
        '--on-primary': '#B4B4B4',
        '--on-primary-active': '#F2F2F2',
        '--frame': '#1E1E1E',
        '--on-frame': '#808080',
        '--on-frame-active': 'var(--on-primary)',
      },
      '[data-theme="light"]': {
        '--primary': '#FFFFFF',
        '--on-primary': '#888888',
        '--on-primary-active': '#333333',
        '--frame': '#EFEFEF',
        '--on-frame': '#808080',
        '--on-frame-active': 'var(--on-primary)',
      },
    };
    processor.addComponents(components);
    const css = processor.preflight('test', false, false, true).build();
    expect(css).toMatchSnapshot('css');
  });

  it('addBase', () => {
    expect(processor.addBase({
      'h1': { fontSize: (processor.theme('fontSize.2xl') as FontSize)[0] ?? '1.5rem' },
      'h2': { fontSize: (processor.theme('fontSize.xl') as FontSize)[0] ?? '1.25rem' },
      'h3': { fontSize: (processor.theme('fontSize.lg') as FontSize)[0] ?? '1.125rem' },
    }).map(i=>i.build()).join('\n')).toBe('h1 {\n  font-size: 1.5rem;\n}\nh2 {\n  font-size: 1.25rem;\n}\nh3 {\n  font-size: 1.125rem;\n}');
  });

  it('add static base styles', () => {
    const processor = new Processor();
    processor.addBase({
      'h1': { fontSize: (processor.theme('fontSize.2xl') as FontSize)[0] ?? '1.5rem' },
      'h2': { fontSize: (processor.theme('fontSize.xl') as FontSize)[0] ?? '1.25rem' },
    }, false);

    expect(processor.preflight('<h1>Hello World</h1>', false, false, true).build()).toEqual('h1 {\n  font-size: 1.5rem;\n}\nh2 {\n  font-size: 1.25rem;\n}');
  });

  it('addVariant pseudoClass', () => {
    const test = new Style('.float-right', new Property('float', 'right'));
    const style = processor.addVariant('disable', ({ pseudoClass }) => {
      return pseudoClass('disabled');
    });
    expect(Array.isArray(style) || style.extend(test).build()).toMatchSnapshot('extend');
    expect(processor.interpret('disable:float-right').styleSheet.build()).toMatchSnapshot('float');
  });

  it('addVariant modifySelectors', () => {
    const test = new Style('.float-right', new Property('float', 'right'));
    const style = processor.addVariant('disabled', ({ modifySelectors, separator }) => {
      return modifySelectors(({ className }) => {
        return `.${processor.e(`disabled${separator}${className}`)}:disabled`;
      });
    });
    const css = Array.isArray(style) || style.extend(test).build();
    expect(css).toMatchSnapshot('css');
  });

  it('#72 plugin test', () => {
    const processor = new Processor({
      plugins: [
        {
          handler: ({ addUtilities }) => {
            addUtilities({});
          },
          config: {
            theme: {
              extend: {
                colors: {
                  active: 'var(--active)',
                  primary: 'var(--primary)',
                  'on-primary': 'var(--on-primary)',
                  'on-primary-active': 'var(--on-primary-active)',
                  frame: 'var(--frame)',
                  'on-frame': 'var(--on-frame)',
                  'on-frame-active': 'var(--on-frame-active)',
                },
              },
            },
          },
        },
      ],
    });
    expect(processor.theme('colors.white')).toEqual('#fff');
    expect(processor.theme('colors.active')).toEqual('var(--active)');
    expect(processor.interpret('bg-active bg-on-primary-active').styleSheet.build()).toEqual(`.bg-active {
  background-color: var(--active);
}
.bg-on-primary-active {
  background-color: var(--on-primary-active);
}`);
  });
});
