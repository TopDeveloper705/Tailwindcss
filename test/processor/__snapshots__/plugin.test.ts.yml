Tools / add duplicated components / css / 0: |-
  .btn {
    padding: .5rem 1rem;
    border-radius: .25rem;
    font-weight: 600;
  }
Tools / add duplicated utilities / css / 0: |-
  .btn {
    padding: .5rem 1rem;
    border-radius: .25rem;
    font-weight: 600;
  }
Tools / add multi utilities / a / 1: |-
  .a {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
Tools / add multi utilities / b / 2: |-
  .b:hover {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
Tools / add multi utilities / preflight / 0: |-
  [type='button'] {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
Tools / add nest class name / css / 0: |-
  .btn.loading:before {
    border-radius: 9999px;
    border-width: 2px;
    height: 1rem;
    margin-right: .5rem;
    width: 1rem;
    -webkit-animation: spin 2s linear infinite;
    animation: spin 2s linear infinite;
    content: "";
    border-color: transparent currentColor currentColor transparent;
  }
Tools / add pseudo components / css / 0: |-
  .btn {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
  .btn:hover {
    color: #fff;
  }
  @media (min-width: 768px) {
    @keyframes ping {
      .btn {
        font-weight: 600;
      }
    }
  }
Tools / add pseudo utilities / css / 0: |-
  .btn {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
  .btn:hover {
    color: #fff;
  }
  @media (min-width: 768px) {
    @keyframes ping {
      .btn {
        font-weight: 600;
      }
    }
  }
  .hover\:abc:hover {
    color: #1c1c1e;
  }
  @media (min-width: 640px) {
    .hover\:abc:hover {
      color: #fff;
    }
  }
  @media (min-width: 768px) {
    @keyframes ping {
      .test {
        color: red;
      }
    }
  }
'Tools / addComponents #242 / css / 0': |-
  .avatar>div {
    display: block;
    overflow: hidden;
  }
  .avatar+div {
    display: block;
    overflow: hidden;
  }
  .avatar div {
    display: block;
    overflow: hidden;
  }
  .avatar~div {
    display: block;
    overflow: hidden;
  }
Tools / addComponents with non-class styles  / css / 0: |-
  [data-theme] {
    --active: #0099FF;
  }
  [data-theme="dark"] {
    --primary: #111111;
    --on-primary: #B4B4B4;
    --on-primary-active: #F2F2F2;
    --frame: #1E1E1E;
    --on-frame: #808080;
    --on-frame-active: var(--on-primary);
  }
  [data-theme="light"] {
    --primary: #FFFFFF;
    --on-primary: #888888;
    --on-primary-active: #333333;
    --frame: #EFEFEF;
    --on-frame: #808080;
    --on-frame-active: var(--on-primary);
  }
Tools / addVariant modifySelectors / css / 0: |-
  .disabled\:float-right:disabled {
    float: right;
  }
Tools / addVariant pseudoClass / css / 1: |-
  .hocus\:float-right:hover:focus {
    float: right;
  }
Tools / addVariant pseudoClass / extend / 0: |-
  .float-right:hover:focus {
    float: right;
  }
Tools / dump config / css / 0: |-
  .bg-nord0 {
    --tw-bg-opacity: 1;
    background-color: rgba(46, 52, 64, var(--tw-bg-opacity));
  }
